smash/factory/net/_loss.py

```diff
@@ -24,106 +25,63 @@ def _hcost(instance: Model) -> float:
-
-        elif name in instance.rr_initial_states.keys:
-            ind = np.argwhere(instance.rr_initial_states.keys == name).item()
+    # % Update rr_parameters and/or rr_initial_states using net
+    y_reshaped = _net_to_parameters(net, x, model_params_states, parameters)
 
-            instance.rr_initial_states.values[..., ind] = y[..., i]
-
-        else:  # nn_parameters excluded from descriptors-to-parameters mapping
-            pass
+    # % Change the mapping to trigger distributed control to get distributed gradients
+    wrap_options.optimize.mapping = "distributed"
 
     # % Run adjoint model
     wrap_parameters_to_control(
         instance.setup,
         instance.mesh,
         instance._input_data,
-        instance._parameters,
+        parameters,
         wrap_options,
     )
-
-    parameters_b = instance._parameters.copy()
-    output_b = instance._output.copy()
-    output_b.cost = np.float32(1)
-
-    wrap_forward_run_b(
-        instance.setup,
-        instance.mesh,
-        instance._input_data,
-        instance._parameters,
-        parameters_b,
-        instance._output,
-        output_b,
-        wrap_options,
-        wrap_returns,
-    )
+    parameters_b = _get_parameters_b(instance, parameters, wrap_options, wrap_returns)
 
     wrap_control_to_parameters(
         instance.setup, instance.mesh, instance._input_data, parameters_b, wrap_options
     )
 
+    # % Reset mapping to ann
+    wrap_options.optimize.mapping = "ann"
+
     # % Get the gradient of regionalization NN
     grad_reg = []
-    for name in parameters:
-        if name in instance.rr_parameters.keys:
-            ind = np.argwhere(instance.rr_parameters.keys == name).item()
+    for name in model_params_states:
+        if name in parameters.rr_parameters.keys:
+            ind = np.argwhere(parameters.rr_parameters.keys == name).item()
 
             grad_reg.append(parameters_b.rr_parameters.values[..., ind])
 
-        elif name in instance.rr_initial_states.keys:
-            ind = np.argwhere(instance.rr_initial_states.keys == name).item()
+        elif name in parameters.rr_initial_states.keys:
+            ind = np.argwhere(parameters.rr_initial_states.keys == name).item()
 
             grad_reg.append(parameters_b.rr_initial_states.values[..., ind])
 
         else:  # nn_parameters excluded from descriptors-to-parameters mapping
             pass
 
-    if return_3d_grad:  # in case of CNN
-        grad_reg = np.transpose(grad_reg, (1, 2, 0))
-    else:
+    if y_reshaped:  # in case of Dense (MLP)
         grad_reg = np.reshape(grad_reg, (len(grad_reg), -1)).T
 
+    else:  # in case of CNN
+        grad_reg = np.transpose(grad_reg, (1, 2, 0))
+
     # % Get the gradient of parameterization NN if used
     grad_par = [
         getattr(parameters_b.nn_parameters, key).copy()
         for key in OPTIMIZABLE_NN_PARAMETERS[max(0, instance.setup.n_layers - 1)]
     ]
 
     return grad_reg, grad_par
```

Comment:
pareil est ce que reg peut etre remplacer par d2p ? En mettant un commentaire a un endroit pour dire que d2p ca veut dire descriptor to parameters ?

Reply:
Dis moi si je me trompe mais grad_reg, c'est le dJ/dtheta en full distrib de rr_parameters et rr_initial_states et grad_par uniquement nn_parameters.

Reply:
oui c'est ça